File: app/src/main/java/com/arjanvlek/oxygenupdater/contribution/SubmittedUpdateFileRepository.java
package com.arjanvlek.oxygenupdater.contribution;

import android.content.ContentValues;
import android.content.Context;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import org.joda.time.DateTimeZone;
import org.joda.time.LocalDateTime;
import org.joda.time.format.ISODateTimeFormat;


/**
 * Oxygen Updater, copyright 2019 Arjan Vlek. File created by arjan.vlek on 02/05/2019.
 */
@SuppressWarnings("WeakerAccess")
public class SubmittedUpdateFileRepository extends SQLiteOpenHelper {

	// If you change the database schema, you must increment the database version.
	private static final int DATABASE_VERSION = 1;
	private static final String DATABASE_NAME = "SubmittedUpdateFiles.db";
	// Table name
	private static final String TABLE_NAME = "news_item";
	// Table columns (id, name, date_submitted)
	private static final String COLUMN_ID = "id";
	private static final String COLUMN_NAME = "name";
	private static final String COLUMN_DATE_SUBMITTED = "date_submitted";
	// Create database
	private static final String SQL_CREATE_TABLE =
			"CREATE TABLE " + TABLE_NAME + " (" +
					COLUMN_ID + " INTEGER PRIMARY KEY," +
					COLUMN_NAME + " TEXT," +
					COLUMN_DATE_SUBMITTED + " TEXT" + // ISO8601 date
					")";
	private SQLiteDatabase readableSqLiteDatabase;
	private SQLiteDatabase writableSqLiteDatabase;


	public SubmittedUpdateFileRepository(Context context) {
		super(context, DATABASE_NAME, null, DATABASE_VERSION);
	}

	@Override
	public SQLiteDatabase getWritableDatabase() {
		if (this.writableSqLiteDatabase == null) {
			this.writableSqLiteDatabase = super.getWritableDatabase();
		}

		return this.writableSqLiteDatabase;
	}

	@Override
	public SQLiteDatabase getReadableDatabase() {
		if (this.readableSqLiteDatabase == null) {
			this.readableSqLiteDatabase = super.getReadableDatabase();
		}

		return this.readableSqLiteDatabase;
	}

	public void onCreate(SQLiteDatabase db) {
		// create DATABASE
		db.execSQL(SQL_CREATE_TABLE);
	}

	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
		// Add code if the database needs to be upgraded.
	}

	public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
		// Add code if the database needs to be downgraded.
	}

	public boolean isFileAlreadySubmitted(String filename) {
		if (filename == null || filename.isEmpty()) {
			return false;
		}

		return DatabaseUtils.queryNumEntries(
				getReadableDatabase(),
				TABLE_NAME,
				COLUMN_NAME + "= ?",
				new String[]{filename}
		) > 0;
	}

	public void store(String filename) {
		if (filename != null && !filename.isEmpty()) {
			ContentValues values = new ContentValues();
			values.put(COLUMN_NAME, filename);
			values.put(COLUMN_DATE_SUBMITTED, LocalDateTime.now(DateTimeZone.forID("Europe/Amsterdam"))
					.toString(ISODateTimeFormat.basicDateTime()));

			getWritableDatabase().insert(TABLE_NAME, null, values);
		}
	}
}

File: app/src/main/java/com/oxygenupdater/database/SubmittedUpdateFileRepository.kt
package com.oxygenupdater.database

import android.content.Context
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.content.contentValuesOf
import com.oxygenupdater.utils.Utils.SERVER_TIME_ZONE
import org.threeten.bp.LocalDateTime

/**
 * @author [Adhiraj Singh Chauhan](https://github.com/adhirajsinghchauhan)
 * @author [Arjan Vlek](https://github.com/arjanvlek)
 */
class SubmittedUpdateFileRepository(context: Context?) : SQLiteOpenHelper(
    context,
    DATABASE_NAME,
    null,
    DATABASE_VERSION
) {

    private var readableSqLiteDatabase: SQLiteDatabase? = null
    private var writableSqLiteDatabase: SQLiteDatabase? = null

    override fun getWritableDatabase(): SQLiteDatabase {
        if (writableSqLiteDatabase == null) {
            writableSqLiteDatabase = super.getWritableDatabase()
        }

        return writableSqLiteDatabase!!
    }

    override fun getReadableDatabase(): SQLiteDatabase {
        if (readableSqLiteDatabase == null) {
            readableSqLiteDatabase = super.getReadableDatabase()
        }

        return readableSqLiteDatabase!!
    }

    override fun onCreate(db: SQLiteDatabase) = db.execSQL(SQL_CREATE_TABLE)

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Add code if the database needs to be upgraded.
    }

    override fun onDowngrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Add code if the database needs to be downgraded.
    }

    fun isFileAlreadySubmitted(filename: String?) = if (filename.isNullOrEmpty()) {
        false
    } else {
        DatabaseUtils.queryNumEntries(
            readableDatabase,
            TABLE_NAME, "$COLUMN_NAME= ?", arrayOf(filename)
        ) > 0
    }

    fun store(filename: String?) {
        if (!filename.isNullOrEmpty()) {
            writableDatabase.insert(
                TABLE_NAME, null, contentValuesOf(
                    COLUMN_NAME to filename,
                    COLUMN_DATE_SUBMITTED to LocalDateTime.now(SERVER_TIME_ZONE).toString()
                )
            )
        }
    }

    companion object {
        // If you change the database schema, you must increment the database version.
        private const val DATABASE_VERSION = 1
        private const val DATABASE_NAME = "SubmittedUpdateFiles.db"

        // Table name
        private const val TABLE_NAME = "news_item"

        // Table columns (id, name, date_submitted)
        private const val COLUMN_ID = "id"
        private const val COLUMN_NAME = "name"
        private const val COLUMN_DATE_SUBMITTED = "date_submitted"

        // Create database
        private const val SQL_CREATE_TABLE = "CREATE TABLE " + TABLE_NAME + " (" +
                COLUMN_ID + " INTEGER PRIMARY KEY," +
                COLUMN_NAME + " TEXT," +
                COLUMN_DATE_SUBMITTED + " TEXT" +  // ISO8601 date
                ")"
    }
}
